src/platform_impl/web/web_sys/canvas/pointer_handler.rs

```diff
@@ -105,11 +115,79 @@ impl PointerHandler {
         ));
     }
 
+    pub fn on_touch_move<F>(&mut self, canvas_common: &super::Common, handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        self.on_touch_move =
+            Some(canvas_common.add_event("pointermove", touch_handler(handler, canvas_common)));
+    }
+
+    pub fn on_touch_down<F>(&mut self, canvas_common: &super::Common, handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        self.on_touch_down =
+            Some(canvas_common.add_event("pointerdown", touch_handler(handler, canvas_common)));
+    }
+
+    pub fn on_touch_up<F>(&mut self, canvas_common: &super::Common, handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        self.on_touch_up =
+            Some(canvas_common.add_event("pointerup", touch_handler(handler, canvas_common)));
+    }
+
+    pub fn on_touch_cancel<F>(&mut self, canvas_common: &super::Common, handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        self.on_touch_cancel =
+            Some(canvas_common.add_event("pointercancel", touch_handler(handler, canvas_common)));
+    }
+
     pub fn remove_listeners(&mut self) {
         self.on_cursor_leave = None;
         self.on_cursor_enter = None;
         self.on_cursor_move = None;
         self.on_pointer_press = None;
         self.on_pointer_release = None;
+        self.on_touch_move = None;
+        self.on_touch_down = None;
+        self.on_touch_up = None;
+        self.on_touch_cancel = None;
+    }
+}
+
+fn touch_handler<F>(mut handler: F, canvas_common: &super::Common) -> impl FnMut(PointerEvent)
+where
+    F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+{
+    let canvas = canvas_common.raw.clone();
+    move |event: PointerEvent| {
+        if event.pointer_type() != "touch" {
+            return;
+        }
+
+        handler(
+            event.pointer_id(),
+            touch_physical_position(&event, &canvas),
+            Force::Normalized(event.pressure() as f64),
+        );
+    }
+}
+
+fn touch_physical_position(
+    event: &PointerEvent,
+    canvas: &HtmlCanvasElement,
+) -> PhysicalPosition<f64> {
+    // Calculate the scale factor every time, since it's possible for the canvas's width and height attributes to get out of sync with its actual HTML size.
+    let dpi_width = canvas.width() as f64 / canvas.offset_width() as f64;
+    let dpi_height = canvas.height() as f64 / canvas.offset_height() as f64;
+    PhysicalPosition {
+        x: event.client_x() as f64 * dpi_width,
+        // Flip the Y axis because canvas's origin is top-left.
+        y: canvas.height() as f64 - event.client_y() as f64 * dpi_height,
     }
 }
```

Comment:
We should probably handle the canvas being resized, but I don't believe we do that anywhere right now (see #1661). However, I'm not sure if this is the correct approach, and I'd prefer such a change to be wholistic rather than touching only one part of the API. Is touch input somehow completely borked if you don't do this?

Reply:
There are cases where resolution changes and resizes are not synchronized, and I had that problem when I developed a game https://github.com/Hihaheho/Hug using this changeset. However, this is a rare case and may not need to be resolved in this PR.

Reply:
I'm starting to remember. A case I struggled with and resolved is scale factor change. In the Hug game, I should change the scale factor for iOS devices for performance reasons, but the change is not reflected in `scale_factor`.

Reply:
https://github.com/rust-windowing/winit/blob/master/src/platform_impl/web/web_sys/mod.rs#L79-L82
This is the source of the problem. `scale_factor` of the web platform always returns `device_pixel_ratio`.

Reply:
> I'm starting to remember. A case I struggled with and resolved is scale factor change. In the Hug game, I should change the scale factor for iOS devices for performance reasons, but the change is not reflected in `scale_factor`.

Not sure if I follow... How would you "change the scale factor?". Is there some web API or winit api to override this? Is it if you transform the canvas html element somehow (with css transforms and such?)

@maroider @ryo33  In the interest of getting this merged. How should we move forward? Some options I could think of:

1. Rename `touch_physical_position` to something non-touch specific and also call it form the the pointer events as well?
2. Move the workaround to `web::Window::scale_factor` and check the canvas `offset_width`/height there. Should probably also get rid of `web_sys::scale_factor` or perhaps rename it to `device_pixel_ratio` since it's apparantly misleading?
3. Remove the workaround from this PR, and deal with the issue separately?

Reply:
I'd prefer to deal with the issue separately.